### PR TITLE
refactor: prereqs for puya debugging support

### PR DESCRIPTION
**Pull request by aorumbayev**
## Proposed Changes

- Deprecating `persistSourceMaps`
- Additionally addresses issues with node specific debug utils, making utils-ts isomorphic and splitting node specific debug utils into a separate package `algokit-utils-debug`

## Further Notes

Just a draft - can be used with per-branch installed as part of related work streams on puya debugging support

### CODE REVIEW

**Comment by aorumbayev** [comment]
@neilcampbell merged your spike and removed all usages of node specific utils. Have a few options to discuss around testing the debug package from utils-ts side, let's sync up on this tomorrow.

**Comment by aorumbayev** [comment]
@neilcampbell @robdmoore last couple of commits also add the following changes:
- deprecating makerx/tsconfig in favour of tsconfig/node20
- disabling console logs on jest which made it hard to navigate/monitor the ci runs -> when an error actually happens though it will print out stderr

File: src/account/account.ts [hunk 1]
```diff
@@ -129,7 +129,7 @@ export function getAccountAddressAsString(addressEncodedInB64: string): string {
   return algosdk.encodeAddress(Buffer.from(addressEncodedInB64, 'base64'))
 }
 
-type NumberConverter<T extends AccountInformationModel> = { [key in keyof T]: ToNumberIfExtends<T[key], number | bigint> }
+export type NumberConverter<T extends AccountInformationModel> = { [key in keyof T]: ToNumberIfExtends<T[key], number | bigint> }
```
Comment (neilcampbell) [inline src/account/account.ts:132]: Is this needed?

File: src/transaction/transaction.ts [hunk 2]
```diff
@@ -622,11 +622,9 @@ export const sendAtomicTransactionComposer = async function (atcSend: AtomicTran
 
     if (Config.debug && Config.projectRoot && Config.traceAll) {
```
Comment (neilcampbell) [inline src/transaction/transaction.ts:623]: Should we still be checking for `Config.projectRoot`? If I understand correctly, it won't be set by us anymore.

File: src/transaction/transaction.ts [hunk 3]
```diff
@@ -680,19 +678,11 @@ export const sendAtomicTransactionComposer = async function (atcSend: AtomicTran
         'Received error executing Atomic Transaction Composer and debug flag enabled; attempting simulation to get more information',
         e,
       )
-      let simulate = undefined
+      const simulate = await performAtomicTransactionComposerSimulate(atc, algod)
       if (Config.debug && Config.projectRoot && !Config.traceAll) {
```
Comment (neilcampbell) [inline src/transaction/transaction.ts:682]: Same comment about not needing `Config.projectRoot`

File: src/types/app-client.ts [hunk 4]
```diff
@@ -192,8 +191,7 @@ export interface AppClientDeployParams extends AppClientDeployCoreParams, AppCli
   schema?: Partial<AppStorageSchema>
 }
 
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-export interface AppClientCallRawArgs extends RawAppCallArgs {}
+export type AppClientCallRawArgs = RawAppCallArgs
```
Comment (neilcampbell) [inline src/types/app-client.ts:194]: I'm wondering if we revert this back, so we're consistently exporting interfaces.

File: src/types/app-client.ts [hunk 5]
```diff
@@ -1435,14 +1430,11 @@ export class ApplicationClient {
     this._clearSourceMap = clearCompiled?.sourceMap
 
     if (Config.debug && Config.projectRoot) {
```
Comment (neilcampbell) [inline src/types/app-client.ts:1432]: Same comment about not needing `Config.projectRoot`

File: src/types/app-factory.ts [hunk 6]
```diff
@@ -458,6 +460,16 @@ export class AppFactory {
       this._clearSourceMap = result.compiledClear.sourceMap
     }
 
+    // Add source map persistence logic
+    if (Config.debug && Config.projectRoot) {
```
Comment (neilcampbell) [inline src/types/app-factory.ts:464]: Same comment about not needing `Config.projectRoot`

Comment (neilcampbell) [reply src/types/app-factory.ts:464]: Actually this function appears to call `AppClient.compile`. If i'm tracing things through correctly won't this even be emitted from there?

File: src/types/app.ts [hunk 7]
```diff
@@ -229,8 +229,7 @@ export interface AppCallTransactionResultOfType<T> extends SendTransactionResult
 }
 
 /** Result from calling an app */
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-export interface AppCallTransactionResult extends AppCallTransactionResultOfType<ABIReturn> {}
+export type AppCallTransactionResult = AppCallTransactionResultOfType<ABIReturn>
```
Comment (neilcampbell) [inline src/types/app.ts:232]: I think we revert this.

File: src/types/composer.ts [hunk 8]
```diff
@@ -1255,11 +1256,9 @@ export default class AlgoKitComposer {
       // Dump the traces to a file for use with AlgoKit AVM debugger
```
Comment (neilcampbell) [inline src/types/composer.ts:1256]: Same comment about not needing `Config.projectRoot`

File: src/types/composer.ts [hunk 9]
```diff
@@ -431,7 +432,7 @@ export type AppMethodCall<T> = Expand<Omit<T, 'args'>> & {
   )[]
 }
 
-type Txn =
+export type Txn =
```
Comment (neilcampbell) [inline src/types/composer.ts:435]: Is this export needed?

File: src/types/debugging.ts [hunk 10]
```diff
@@ -1,166 +1,41 @@
...
-   * Creates a source map from a dictionary of source map data.
-   * @param data The data
-   * @returns The source map
-   */
-  public static fromDict(data: AVMDebuggerSourceMapDict): AVMDebuggerSourceMap {
-    return new AVMDebuggerSourceMap(
-      data['txn-group-sources'].map((item) => new AVMDebuggerSourceMapEntry(item['sourcemap-location'], item['hash'])),
-    )
-  }
+// change event listener https://github.com/algorandfoundation/algokit-subscriber-ts/blob/main/src/subscriber.ts#L259
```
Comment (neilcampbell) [inline src/types/debugging.ts:8]:
```suggestion
```

File: src/types/debugging.ts [hunk 11]
```diff
@@ -1,166 +1,41 @@
...
-    if (fileName.endsWith('.teal')) {
-      return fileName.slice(0, -5)
-    }
-    return fileName
-  }
 }
 
-/** Parameters to a call that persists source maps */
-export interface PersistSourceMapsParams {
+/**
```
Comment (neilcampbell) [inline src/types/debugging.ts:27]: All these jsdocs should be reviewed and tweaked.

File: src/account/account.ts (hunk 1 above)
Comment (aorumbayev) [reply src/account/account.ts:132]: a fix for a typedoc warning, this is referenced in public facing abstraction but not explicitly exported

File: src/types/composer.ts (hunk 9 above)
Comment (aorumbayev) [reply src/types/composer.ts:435]: a fix for a typedoc warning, this is referenced in public facing abstraction but not explicitly exported

File: docs/capabilities/event-emitter.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
+# Event Emitter
+
+The Event Emitter is a capability provided by AlgoKit Utils that allows for asynchronous event handling. It provides a flexible mechanism for emitting and listening to custom events, which can be particularly useful for debugging and extending functionality not available in the `algokit-utils-ts` package.
```
Comment (neilcampbell) [inline docs/capabilities/event-emitter.md:3]:
```suggestion
The Event Emitter is a capability provided by AlgoKit Utils that allows for asynchronous event handling of lifecycle events. It provides a flexible mechanism for emitting and listening to custom events, which can be particularly useful for debugging and extending functionality not available in the `algokit-utils-ts` package.
```
My mental model for this is that we have a mechanism for emitting lifecycle type events that users can choose to handle.

File: docs/capabilities/event-emitter.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+```typescript
+enum EventType {
+  TxnGroupSimulated = 'TxnGroupSimulated',
+  AppCompiled = 'AppCompiled',
+}
+```
+
+These events are primarily used for debugging purposes and are handled by the [`algokit-utils-ts-debug`](https://github.com/algorandfoundation/algokit-utils-ts-debug) package. However, the `AsyncEventEmitter` is designed to be generic and can be extended to handle custom event types as well.
+
+## Emitting Events
```
Comment (neilcampbell) [inline docs/capabilities/event-emitter.md:30]: The emitting side is mostly an internal concern for the utils library, so I'd remove this section and wouldn't focus on that side of things.

File: src/types/async-event-emitter.spec.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import { algoKitLogCaptureFixture } from '../testing/fixtures/algokit-log-capture-fixture'
+
+describe('async-event-emitter', () => {
+  const logging = algoKitLogCaptureFixture()
+  beforeEach(logging.beforeEach)
+  afterEach(logging.afterEach)
+
+  const mockRegisterNodeDebugHandlers = () => {
+    // TODO: Define types for the events
+    // TODO: Question - Would this be the package we'd use to surface the UI widget for downloading a simulate trace?
```
Comment (neilcampbell) [inline src/types/async-event-emitter.spec.ts:12]:
```suggestion
```

File: src/types/config.ts [hunk 15]
```diff
@@ -136,3 +112,6 @@ export class UpdatableConfig implements Readonly<Config> {
     this.config = { ...this.config, ...newConfig }
   }
 }
+
```
Comment (neilcampbell) [inline src/types/config.ts:115]: What the context of adding this?

Comment (aorumbayev) [reply src/types/config.ts:115]: No context, shouldn't be there - removed

**Review by neilcampbell — APPROVED** [review]

**Pull request merged**